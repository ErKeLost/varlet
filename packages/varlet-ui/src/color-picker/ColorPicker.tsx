import { defineComponent, ref, toRefs, onMounted, watch, computed } from 'vue'
import { colorPickerProps, ColorPickerProps } from './props'
import { createNamespace, call } from '../utils/components'
import { parseBaseColor, extractBaseColor, HSVtoCSS, nullColor } from './utils/color-utils'
import { HSV } from './utils/color-utils-types'
import VarColorPickerCanvas from './components/color-picker-canvas/color-picker-canvas'
import VarColorPickerEdit from './components/color-picker-edit/color-picker-edit'
import VarColorPickerSwatches from './components/color-picker-swatches/color-picker-swatches'
import VarColorPickerHueSlider from './components/color-picker-hue-slider/color-picker-hue-slider'
import VarColorPickerAlphaSlider from './components/color-picker-alpha-slider/color-picker-alpha-slider'
import './colorPicker.less'
import '../styles/elevation.less'

const DEFAULT_MODE = 'rgba'
export default defineComponent({
  name: 'VarColorPicker',
  props: colorPickerProps,
  setup(props: ColorPickerProps) {
    const { n, classes } = createNamespace('color-picker')

    const { modelValue, mode, disabled, modes } = toRefs(props)
    const currentMode = ref(DEFAULT_MODE)
    const dragTransferFlag = ref<boolean>(false)
    const initialColor = ref<any>()

    function updateModelValueColor(color: any) {
      initialColor.value = parseBaseColor(color) ?? nullColor
      const value = extractBaseColor(parseBaseColor(color) ?? nullColor, props.modelValue)
      call(props['onUpdate:modelValue'], value)
    }

    function updateColor(hsva: HSV) {
      updateModelValueColor(hsva)
    }

    function updateMode(mode: string) {
      currentMode.value = mode
    }

    function handleDragger(flag: boolean) {
      dragTransferFlag.value = flag
    }

    const containerStyle = computed(() => {
      return {
        width: `${props.width}px`,
        maxWidth: `${props.width}px`,
      }
    })

    const dotStyle = computed(() => {
      const defaultColor = HSVtoCSS(initialColor.value ?? nullColor)
      return {
        background: defaultColor,
        border: `1px solid ${defaultColor}`,
      }
    })

    const dotContainerStyle = computed(() => {
      const defaultColor = HSVtoCSS(initialColor.value ?? nullColor)
      return {
        border: `14px solid ${defaultColor}`,
      }
    })

    onMounted(() => {
      if (!props.modes.includes(mode.value)) mode.value = props.modes[0]
    })

    watch(
      () => modelValue.value,
      (newV, oldV) => {
        if (oldV !== newV && !dragTransferFlag.value) {
          updateModelValueColor(newV)
        }
      },
      {
        immediate: true,
      }
    )

    watch(
      () => mode.value,
      (newV) => {
        updateMode(newV)
      }
    )

    return () => {
      return (
        <>
          <div class={classes(n(), n('$-elevation--2'))} style={containerStyle.value}>
            <div></div>
            {props.canvasLayout && (
              <VarColorPickerCanvas
                width={props.width}
                height={props.canvasHeight}
                color={initialColor.value}
                onUpdate:color={updateColor}
                dragger={handleDragger}
                disabled={disabled.value}
              />
            )}
            {(props.sliderLayout || props.inputLayout) && (
              <div class={n('control')}>
                {props.sliderLayout && (
                  <div class={n('preview')}>
                    <div class={n('preview__dots')} style={dotContainerStyle.value}>
                      <div style={dotStyle.value} />
                    </div>
                    <div class={n('preview__slider')}>
                      <VarColorPickerHueSlider
                        color={initialColor.value}
                        onUpdate:color={updateColor}
                        disabled={disabled.value}
                      />

                      {currentMode.value.endsWith('a') ? (
                        <VarColorPickerAlphaSlider
                          color={initialColor.value}
                          onUpdate:color={updateColor}
                          disabled={disabled.value}
                        />
                      ) : null}
                    </div>
                  </div>
                )}
                {props.inputLayout && (
                  <VarColorPickerEdit
                    color={initialColor.value}
                    onUpdate:color={updateColor}
                    disabled={disabled.value}
                    modes={modes.value}
                    mode={currentMode.value}
                    onUpdate:mode={updateMode}
                  />
                )}
              </div>
            )}
            {props.swatchesLayout && (
              <VarColorPickerSwatches
                color={initialColor.value}
                onUpdate:color={updateColor}
                disabled={disabled.value}
              />
            )}
          </div>
        </>
      )
    }
  },
})
