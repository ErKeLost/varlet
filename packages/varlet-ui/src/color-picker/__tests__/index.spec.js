import ColorPicker from '..'
import { createApp } from 'vue'
import { mount } from '@vue/test-utils'

test('test color-picker plugin', () => {
  const app = createApp({}).use(ColorPicker)
  expect(app.component(ColorPicker.name)).toBeTruthy()
})

test('test color-picker snapshot', () => {
  const wrapper = mount(ColorPicker)
  expect(wrapper.html()).toMatchSnapshot()
})
