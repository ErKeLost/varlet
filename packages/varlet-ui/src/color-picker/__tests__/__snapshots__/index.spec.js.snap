// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test color-picker snapshot 1`] = `
"<div class="var-color-picker var-elevation--2" style="width: 300px; max-width: 300px;">
  <div></div>
  <div class="var-color-picker-canvas" style="width: 300px; height: 150px; cursor: default;"><canvas></canvas>
    <div style="top: 0px; left: 0px;" class="var-color-picker-canvas__dot"></div>
  </div>
  <div class="var-color-picker__control">
    <div class="var-color-picker__preview">
      <div class="var-color-picker__preview__dots">
        <div style="background: rgb(255, 255, 255); border: 1px solid rgba(255, 255, 255, 1);"></div>
      </div>
      <div class="var-color-picker__preview__slider">
        <div class="var-color-picker-hue-slider">
          <div class="var-color-picker-hue-slider__bar">
            <div class="var-color-picker-hue-slider__bar-pointer" style="left: 0px; top: 0px; transition: all 0.3s ease;">
              <div class="var-color-picker-hue-slider__bar-handle"></div>
            </div>
          </div>
        </div>
        <div class="var-color-picker-alpha-slider transparent">
          <div class="var-color-picker-alpha-slider__bar" style="--color-picker-alpha-slider: #FFFFFFFF;">
            <div class="var-color-picker-alpha-slider__bar-pointer" style="left: 0px; top: 0px; transition: all 0.3s ease;">
              <div class="var-color-picker-alpha-slider__bar-handle"></div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="var-color-picker-edit">
      <div class="var-color-picker-edit__input"><input type="number" min="0" max="255" step="1"><span>R</span></div>
      <div class="var-color-picker-edit__input"><input type="number" min="0" max="255" step="1"><span>G</span></div>
      <div class="var-color-picker-edit__input"><input type="number" min="0" max="255" step="1"><span>B</span></div>
      <div class="var-color-picker-edit__input"><input type="number" min="0" max="1" step="0.01"><span>A</span></div>
      <var-button round="true" text="true">
        <div class="var-color-picker-edit__icon">
          <var-icon size="20" name="unfold-more-horizontal"></var-icon>
        </div>
      </var-button>
    </div>
  </div>
  <!---->
</div>"
`;
